src/libappimage/desktop_integration/integrator/Integrator.cpp

```diff
@@ -0,0 +1,267 @@
+// system
+#include <cstdlib>
+#include <sstream>
+#include <utility>
+#include <vector>
+#include <iostream>
+#include <fstream>
+
+// libraries
+#include <boost/filesystem.hpp>
+#include <boost/algorithm/string/trim.hpp>
+#include <boost/algorithm/string/replace.hpp>
+#include <boost/algorithm/string.hpp>
+#include <XdgUtils/DesktopEntry/DesktopEntry.h>
+#include <XdgUtils/BaseDir/BaseDir.h>
+
+// local
+#include "appimage/appimage.h"
+#include "appimage/core/AppImage.h"
+#include "appimage/desktop_integration/exceptions.h"
+#include "utils/HashLib.h"
+#include "utils/IconHandle.h"
+#include "DesktopEntryEditor.h"
+#include "ResourcesExtractor.h"
+#include "Integrator.h"
+
+namespace bf = boost::filesystem;
+
+using namespace appimage::core;
+using namespace XdgUtils::DesktopEntry;
+
+namespace appimage {
+    namespace desktop_integration {
+        namespace integrator {
+            /**
+             * Implementation of the opaque pointer patter for the integrator class
+             * see https://en.wikipedia.org/wiki/Opaque_pointer
+             *
+             * Contain a set of helper methods that will be used at the integrator class to fulfill the different task
+             */
+            struct Integrator::Priv {
+                core::AppImage appImage;
+                bf::path xdgDataHome;
+                std::string appImageId;
+                std::string vendorPrefix = "appimagekit";
+
+                DesktopEntry entry;
+                DesktopIntegrationResources resources;
+
+                Priv(const AppImage& appImage, const std::string& xdgDataHome) : appImage(appImage),
+                                                                                 xdgDataHome(xdgDataHome) {
+
+                    if (xdgDataHome.empty())
+                        Priv::xdgDataHome = XdgUtils::BaseDir::XdgDataHome();
```

Comment:
> `Priv::xdgDataHome` is more efficient than `this->xdgDataHome` it doesn't require any pointer resolution.

Why do you think so? Why should that be more efficient? Sure, at a first glance, it might seem like "pointer access vs. some other way", but in the end the compiler can resolve both ways to the same variables. I _highly_ doubt that either way is "faster". Do you have any docs that show that `this` is slower? I'm curious.

Reply:
As you said if the compiler is smart enough `this->` should be as fast as `Class::`. But why should we depend on the compiler for this?

Reply:
> But why should we depend on the compiler for this?

What do you think a compiler does when you use that syntax? Of course it will use the `this` pointer there, too! It's not static memory or so, it's not magic.

If you don't believe me, there's an ***awesome*** project called https://cppinsights.io, which shows you what clang does internally when interpreting source code.

For instance, I wrote this snippet:

```cpp
#include <cstdio>

class A {
private:
  int a = 0;
  
public:
  int getA() {
    return this->a;
  }

  int getA2() {
    return A::a;
  }
};

int main() {}
```

This is converted into:

```cpp
#include <cstdio>

class A {
private:
  int a = 0;
  
public:
  inline int getA()
  {
    return this->a;
  }
  

  inline int getA2()
  {
    return this->a;
  }
  
};

int main()
{
}
```

Surprise! The `this` pointer is used there, too.

Conclusion: as said previously, this is not an optimization or anything, it's just a different syntax. Accessing the `this` pointer doesn't perform any costly dereferencing or so (unless it's necessary). Omitting `this->` or using `MyClass::` isn't some smart magic syntax that saves a pointer operation.

Reply:
Ok, it seems that they are the same.